python/ray/train/ml_trainer.py

```diff
@@ -0,0 +1,108 @@
+            it will be fit on this dataset and this dataset will be
+            transformed.
+        additional_datasets (Optional[GenDataset]): Any additional
+            Datasets (such as validation or test datasets) to use for
+            training. If a ``preprocessor`` is provided, the datasets
+            specified here will only be transformed, and not fit on.
+        preprocessor (Optional[Preprocessor]): A preprocessor to preprocess
+            the provided datasets.
+        resume_from_checkpoint (Optional[Checkpoint]): A checkpoint to
+            resume training from.
+    """
+
+    def __init__(
+        self,
+        scaling_config: Optional[ScalingConfig] = None,
+        run_config: Optional[RunConfig] = None,
+        train_dataset: Optional[GenDataset] = None,
+        additional_datasets: Optional[Dict[str, GenDataset]] = None,
+        preprocessor: Optional[Preprocessor] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+        self.scaling_config = scaling_config
+        self.run_config = run_config if run_config else {}
+        self.train_dataset = train_dataset
+        self.additional_datasets = additional_datasets
+        self.preprocessor = preprocessor
+        self.checkpoint_to_resume_from = resume_from_checkpoint
+
+    def fit(self) -> Result:
+        """Runs training.
+
+        Returns:
+            A Result object containing the training result.
+        """
+        trainable = self.as_trainable()
+
+        # Copied from initial prototyping.
+        # TODO: Replace with Tuner.
+        analysis = tune.run(run_or_experiment=trainable, **self.run_config)
+
+        assert len(analysis.trials) == 1
+
+        trial = analysis.trials[0]
+
+        result = Result(
+            metrics=trial.last_result,
+            checkpoint=Checkpoint.from_directory(trial.checkpoint.value)
+            if trial.checkpoint.value
+            else None,
+        )
+
+        return result
+
+    def _override_attributes_with_config(self, config: Dict):
+        """Overrides attributes of the Trainer with values in ``config``.
+
+        This is needed to allow attributes of Trainer to be tuned as
+        hyperparameters.
+
+        This method does a deep update of dictionaries.
+
+        Args:
+            config (Dict): A dictionary to update attributes with.
+        """
+        for key in list(config.keys()):
+            if hasattr(self, key):
+                current_attribute = getattr(self, key)
+                if isinstance(current_attribute, dict):
+                    # Do a deep update of the dictionary.
+                    deep_update(current_attribute, config[key], new_keys_allowed=True)
+                else:
+                    # Don't do a deep update and directly set the attribute
+                    # to the value in config.
+                    setattr(self, key, config[key])
+                # Remove the key from the dict.
+                del config[key]
```

Comment:
Why mutate the dict?

Reply:
The reasoning was that whatever's "leftover" from the config after overriding attributes are trainer specific configurations. So these can be directly sent to the training function for example.

Though you're right, we shouldn't be mutating the arg itself. Updated the implementation to do a deep copy and then return the leftover key-value pairs.